### PR TITLE
Add a new error for dynamic constant assignment

### LINKED ISSUE #105: Dynamic constant assignment not caught during parsing
https://sorbet.run/#def%20foo%0A%20%20SyntaxErrorInMRIAndInWhiteQuarkParser%20%3D%20100%0Aend

### PR DESCRIPTION
**Pull request by elliottt**
### Motivation

Dynamic constant assignment is a runtime error in ruby, and we currently have partial support for ruling them out. This change forbids them in the context of a method definition.

Fixes #105

### Test plan

See included automated tests.

### CODE REVIEW

File: test/testdata/namer/constant_in_block.rb [hunk 1]
```diff
@@ -1,15 +1,17 @@
...
     C = 1 # this is parse error in Ruby, but not in Sorbet
+  # ^^^^^ error: Dynamic constant assignment
     C # error: Unable to resolve constant
     end
   end
-  
+
   [].each do
     B = 1
+  # ^^^^^ error: Dynamic constant assignment
```
Comment (elliottt) [inline test/testdata/namer/constant_in_block.rb]: I'm on the fence about whether or not this should be an error. I can see not raising an error if the owning context doesn't involve a method definition.

File: namer/namer.cc [hunk 2]
```diff
@@ -618,7 +618,14 @@ class NameInserter {
     }
 
     unique_ptr<ast::Assign> fillAssign(core::MutableContext ctx, unique_ptr<ast::Assign> asgn) {
-        // TODO(nelhage): forbid dynamic constant definition
+
+        // forbid dynamic constant definition
+        if (!ctx.owner.data(ctx)->isClass()) {
+            if (auto e = ctx.state.beginError(asgn->loc, core::errors::Namer::DynamicConstant)) {
```
Comment (DarkDimius) [inline namer/namer.cc]: maybe use different error code?

**Comment by DarkDimius** [comment]
I'm +1 on change,
AFAIK ruby has a similar restriction.
Could you run in on pay-server as an indicator of impact on open-source folks?

File: namer/namer.cc (hunk 2 above)
Comment (elliottt) [reply namer/namer.cc]: Should I make a new one? This one seemed appropriate, and wasn't used anywhere else.

Comment (elliottt) [reply namer/namer.cc]: OK, I made the new code `DynamicConstantAssignment` and switched to that here.

File: test/testdata/namer/constant_in_block.rb (hunk 1 above)
Comment (elliottt) [reply test/testdata/namer/constant_in_block.rb]: It turns out the error added in this PR doesn't get triggered here, as the owning context is the class, not the block.

**Comment by elliottt** [comment]
I've made #1334 to track raising the error in blocks.

**Comment by elliottt** [comment]
This triggers errors for cases like this:
```ruby
class Test
  describe 'foo' do
    it 'bar' do
      DATA = "string"
    end
  end
end
```
as the minitest dsl support turns the `it` blocks into method definitions.

**Comment by elliottt** [comment]
This works fine in pay-server now.

**Comment by DarkDimius** [comment]
I didn't see recent PRs from you landing in pay-server.
Does it mean we didn't introduce any new errors to existing code?
This should be a good indicator of this not breaking open-source.

**Comment by elliottt** [comment]
Yes, I meant that it didn't raise any new errors in pay-server, and didn't require any changes.

**Pull request merged**